Framework/Muon/src/ApplyMuonDetectorGrouping.cpp

```diff
@@ -0,0 +1,503 @@
+#include "MantidMuon/ApplyMuonDetectorGrouping.h"
+#include "MantidMuon/MuonAlgorithmHelper.h"
+
+#include "MantidAPI/AlgorithmManager.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/Workspace.h"
+#include "MantidAPI/WorkspaceGroup.h"
+#include "MantidDataObjects/TableWorkspace.h"
+#include "MantidKernel/EnabledWhenProperty.h"
+#include "MantidKernel/ListValidator.h"
+
+#include <algorithm>
+#include <ctype.h>
+#include <string>
+#include <vector>
+
+const std::vector<std::string> g_analysisTypes = {"Counts", "Asymmetry"};
+
+namespace {
+
+// Convert input string plot type to PlotType.
+Mantid::Muon::PlotType getPlotType(const std::string &plotType) {
+  if (plotType == "Counts") {
+    return Mantid::Muon::PlotType::Counts;
+  } else if (plotType == "Asymmetry") {
+    return Mantid::Muon::PlotType::Asymmetry;
+  } else {
+    // default to Counts.
+    return Mantid::Muon::PlotType::Counts;
+  }
+}
+
+/**
+ * Convert the input workspace into a workspace group if e.g. it has only a
+ * single period otherwise leave it alone.
+ */
+Mantid::API::WorkspaceGroup_sptr
+convertInputWStoWSGroup(Mantid::API::Workspace_sptr inputWS) {
+
+  // Cast input WS to a WorkspaceGroup
+  auto muonWS = boost::make_shared<Mantid::API::WorkspaceGroup>();
+  if (auto test =
+          boost::dynamic_pointer_cast<Mantid::API::MatrixWorkspace>(inputWS)) {
+    muonWS->addWorkspace(test);
+  } else {
+    muonWS = boost::dynamic_pointer_cast<Mantid::API::WorkspaceGroup>(inputWS);
+  }
+  return muonWS;
+}
+
+} // namespace
+
+using namespace Mantid::API;
+using namespace Mantid::DataObjects;
+using namespace Mantid::Kernel;
+using namespace Mantid::MuonAlgorithmHelper;
+
+namespace Mantid {
+namespace Muon {
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(ApplyMuonDetectorGrouping)
+
+void ApplyMuonDetectorGrouping::init() {
+
+  std::string emptyString("");
```

Comment:
I would just use `""` directly in each space. This actually saves the compiler having to copy this each time.

Reply:
For some reason I get a link error (unresolved external symbol) when using "" directly

Reply:
Okay - don't worry then. It's not worth spending too much time on it.